# Cleaning and preparation

## Imports

In [1]:
import pandas as pd
import numpy as np
import folium

## Initial Dataset
### Loading

In [2]:
# import bike_data

bike_data = pd.read_csv("data/sf_2019.csv")

### Description

In [3]:
bike_data.info(show_counts=True)
# outcome null values for station information

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2506983 entries, 0 to 2506982
Data columns (total 8 columns):
 #   Column              Non-Null Count    Dtype  
---  ------              --------------    -----  
 0   start_time          2506983 non-null  object 
 1   end_time            2506983 non-null  object 
 2   start_station_id    2426249 non-null  float64
 3   end_station_id      2424081 non-null  float64
 4   start_station_name  2426850 non-null  object 
 5   end_station_name    2424633 non-null  object 
 6   bike_id             2506983 non-null  int64  
 7   user_type           2506983 non-null  object 
dtypes: float64(2), int64(1), object(5)
memory usage: 153.0+ MB


In [4]:
bike_data.head(5)

start_time             end_time  start_station_id  end_station_id  \
0  2019-01-31 17:57:45  2019-02-01 16:24:50             229.0           196.0   
1  2019-01-31 20:58:34  2019-02-01 15:16:55               4.0           134.0   
2  2019-01-31 18:06:53  2019-02-01 11:30:47             245.0           157.0   
3  2019-01-31 19:46:10  2019-02-01 08:10:51              85.0            53.0   
4  2019-01-31 14:19:02  2019-02-01 07:10:52              16.0            28.0   

            start_station_name              end_station_name  bike_id  \
0    Foothill Blvd at 42nd Ave       Grand Ave at Perkins St     4861   
1  Cyril Magnin St at Ellis St        Valencia St at 24th St     5506   
2       Downtown Berkeley BART          65th St at Hollis St     2717   
3      Church St at Duboce Ave        Grove St at Divisadero     4557   
4      Steuart St at Market St  The Embarcadero at Bryant St     2100   

    user_type  
0  Subscriber  
1  Subscriber  
2    Customer  
3    Customer  
4    Customer

In [5]:
bike_data.tail(5)

start_time             end_time  start_station_id  \
2506978  2019-12-28 18:37:29  2019-12-28 18:49:40               NaN   
2506979  2019-12-28 18:37:24  2019-12-28 18:49:53               NaN   
2506980  2019-12-26 14:25:37  2019-12-26 14:43:02               NaN   
2506981  2019-12-30 20:35:43  2019-12-30 20:41:09               NaN   
2506982  2019-12-31 19:26:48  2019-12-31 20:33:52               NaN   

         end_station_id start_station_name end_station_name  bike_id  \
2506978             NaN                NaN              NaN   719651   
2506979             NaN                NaN              NaN   655411   
2506980             NaN                NaN              NaN   936881   
2506981             NaN                NaN              NaN   397539   
2506982             NaN                NaN              NaN   662304   

          user_type  
2506978  Subscriber  
2506979  Subscriber  
2506980  Subscriber  
2506981  Subscriber  
2506982  Subscriber

In [6]:
bike_data.sample(5)

start_time             end_time  start_station_id  \
111490   2019-01-16 16:54:24  2019-01-16 17:05:16              11.0   
26960    2019-01-29 08:15:28  2019-01-29 08:28:48              30.0   
2113862  2019-10-08 09:42:21  2019-10-08 09:52:27             243.0   
1090322  2019-06-25 07:33:06  2019-06-25 07:37:34              36.0   
926561   2019-05-22 15:58:45  2019-05-22 16:04:57              49.0   

         end_station_id                              start_station_name  \
111490             30.0                          Davis St at Jackson St   
26960              15.0  San Francisco Caltrain (Townsend St at 4th St)   
2113862           244.0                     Bancroft Way at College Ave   
1090322            24.0                             Folsom St at 3rd St   
926561             22.0                             S Park St at 3rd St   

                                          end_station_name  bike_id  \
111490      San Francisco Caltrain (Townsend St at 4th St)     4801   
26960    San Francisco Ferry Building (Harry Bridges Pl...     2293   
2113862                         Shattuck Ave at Hearst Ave       37   
1090322                              Spear St at Folsom St     2702   
926561                               Howard St at Beale St      270   

          user_type  
111490   Subscriber  
26960    Subscriber  
2113862    Customer  
1090322  Subscriber  
926561   Subscriber

Value ranges

In [7]:
print("Range Start time: " + str(bike_data.start_time.min() + " to " + str(bike_data.start_time.max())))
print("Range End time:   " + str(bike_data.end_time.min()  +  " to " + str(bike_data.end_time.max())))

Range Start time: 2019-01-01 00:07:11 to 2019-12-31 23:59:41
Range End time:   2019-01-01 00:10:58 to 2020-01-01 10:37:57


In [8]:
station_ids = bike_data[["start_station_id", "end_station_id"]].dropna().values
station_ids_unique = np.unique(station_ids)

print("Range Station Ids (excl. NaN): " + str(station_ids_unique.min()) + " to " + str(station_ids_unique.max()))
print("Number of unique stations (excl. NaN): " + str(len(station_ids_unique)))

Range Station Ids (excl. NaN): 3.0 to 498.0
Number of unique stations (excl. NaN): 435


In [9]:
print("Range Bike Ids: " + str(bike_data.bike_id.min()) + " to " + str(bike_data.bike_id.max()))
print("Number of unique Bike Ids: " + str(len(np.unique(bike_data.bike_id))))

Range Bike Ids: 4 to 999941
Number of unique Bike Ids: 11070


In [10]:
print("Different user types: " + str(np.unique(bike_data.user_type)))

Different user types: ['Customer' 'Subscriber']


Check for NaN, NULL

In [11]:
bike_data.isnull().values.any()

True

In [12]:
bike_data.info(show_counts=True)

# null values only for "station" columns


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2506983 entries, 0 to 2506982
Data columns (total 8 columns):
 #   Column              Non-Null Count    Dtype  
---  ------              --------------    -----  
 0   start_time          2506983 non-null  object 
 1   end_time            2506983 non-null  object 
 2   start_station_id    2426249 non-null  float64
 3   end_station_id      2424081 non-null  float64
 4   start_station_name  2426850 non-null  object 
 5   end_station_name    2424633 non-null  object 
 6   bike_id             2506983 non-null  int64  
 7   user_type           2506983 non-null  object 
dtypes: float64(2), int64(1), object(5)
memory usage: 153.0+ MB


In [13]:
# count of empty start_station_ids
bike_data.start_station_id.isna().value_counts()

False    2426249
True       80734
Name: start_station_id, dtype: int64

We decided to keep trips with empty start_station_id. 80734 rows are roughly 3 % of the whole dataset and since the prediction is performed to predict the demand of the whole bike fleet regardless of the specific stations, these trips have an influence on the overall demand.

Check for duplicates

In [14]:
bike_data.duplicated().any()

False

Precalculate and add time-based lag features

In [15]:
bike_data['ts_start_hour'] = pd.to_datetime(bike_data.start_time).dt.floor('h')
bike_data['start_hour'] = pd.DatetimeIndex(bike_data.start_time).hour
bike_data['start_weekday'] = pd.DatetimeIndex(bike_data.start_time).weekday
bike_data['start_month'] = pd.DatetimeIndex(bike_data.start_time).month
bike_data['trip_duration'] = (pd.to_datetime(bike_data.end_time) - pd.to_datetime(bike_data.start_time)).dt.total_seconds()

## Joining data with full dataset for geo data

Enhance bike data by adding geo data found on https://www.lyft.com/bikes/bay-wheels/system-data  
Most important, this adds the geo information for stations.

In [16]:
bike_data_full = pd.read_csv("data/Sf_2019_full.csv")
bike_data_full.head(5)

C:\Users\leoni\AppData\Local\Temp\ipykernel_16356\2828992919.py:1: DtypeWarning: Columns (14,15) have mixed types. Specify dtype option on import or set low_memory=False.
  bike_data_full = pd.read_csv("data/Sf_2019_full.csv")


Unnamed: 0  duration_sec                start_time  \
0           0         80825  2019-01-31 17:57:44.6130   
1           1         65900  2019-01-31 20:58:33.8860   
2           2         62633  2019-01-31 18:06:52.9240   
3           3         44680  2019-01-31 19:46:09.7190   
4           4         60709  2019-01-31 14:19:01.5410   

                   end_time  start_station_id           start_station_name  \
0  2019-02-01 16:24:49.8640             229.0    Foothill Blvd at 42nd Ave   
1  2019-02-01 15:16:54.1730               4.0  Cyril Magnin St at Ellis St   
2  2019-02-01 11:30:46.5300             245.0       Downtown Berkeley BART   
3  2019-02-01 08:10:50.3180              85.0      Church St at Duboce Ave   
4  2019-02-01 07:10:51.0650              16.0      Steuart St at Market St   

   start_station_latitude  start_station_longitude  end_station_id  \
0               37.775745              -122.213037           196.0   
1               37.785881              -122.408915           134.0   
2               37.870139              -122.268422           157.0   
3               37.770083              -122.429156            53.0   
4               37.794130              -122.394430            28.0   

               end_station_name  end_station_latitude  end_station_longitude  \
0       Grand Ave at Perkins St             37.808894            -122.256460   
1        Valencia St at 24th St             37.752428            -122.420628   
2          65th St at Hollis St             37.846784            -122.291376   
3        Grove St at Divisadero             37.775946            -122.437777   
4  The Embarcadero at Bryant St             37.787168            -122.388098   

   bike_id   user_type bike_share_for_all_trip rental_access_method  
0     4861  Subscriber                      No                  NaN  
1     5506  Subscriber                      No                  NaN  
2     2717    Customer                      No                  NaN  
3     4557    Customer                      No                  NaN  
4     2100    Customer                      No                  NaN

In [17]:
bike_data_full.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2506983 entries, 0 to 2506982
Data columns (total 16 columns):
 #   Column                   Dtype  
---  ------                   -----  
 0   Unnamed: 0               int64  
 1   duration_sec             int64  
 2   start_time               object 
 3   end_time                 object 
 4   start_station_id         float64
 5   start_station_name       object 
 6   start_station_latitude   float64
 7   start_station_longitude  float64
 8   end_station_id           float64
 9   end_station_name         object 
 10  end_station_latitude     float64
 11  end_station_longitude    float64
 12  bike_id                  int64  
 13  user_type                object 
 14  bike_share_for_all_trip  object 
 15  rental_access_method     object 
dtypes: float64(6), int64(3), object(7)
memory usage: 306.0+ MB


In [18]:
# get df with geo coordinates data for each station
station_geo_data = bike_data_full[['start_station_id', 'start_station_latitude','start_station_longitude']].dropna().drop_duplicates()
station_geo_data.set_index('start_station_id', inplace=True)
station_geo_data

start_station_latitude  start_station_longitude
start_station_id                                                 
229.0                          37.775745              -122.213037
4.0                            37.785881              -122.408915
245.0                          37.870139              -122.268422
85.0                           37.770083              -122.429156
16.0                           37.794130              -122.394430
...                                  ...                      ...
495.0                          37.343208              -121.884390
498.0                          37.725890              -122.452444
229.0                          37.774234              -122.212483
444.0                          37.371116              -121.881332
494.0                          37.326282              -121.904911

[459 rows x 2 columns]

In [19]:
bike_data_joined = bike_data.join(station_geo_data, on='start_station_id')
bike_data_joined

start_time             end_time  start_station_id  \
0        2019-01-31 17:57:45  2019-02-01 16:24:50             229.0   
0        2019-01-31 17:57:45  2019-02-01 16:24:50             229.0   
1        2019-01-31 20:58:34  2019-02-01 15:16:55               4.0   
2        2019-01-31 18:06:53  2019-02-01 11:30:47             245.0   
3        2019-01-31 19:46:10  2019-02-01 08:10:51              85.0   
...                      ...                  ...               ...   
2506978  2019-12-28 18:37:29  2019-12-28 18:49:40               NaN   
2506979  2019-12-28 18:37:24  2019-12-28 18:49:53               NaN   
2506980  2019-12-26 14:25:37  2019-12-26 14:43:02               NaN   
2506981  2019-12-30 20:35:43  2019-12-30 20:41:09               NaN   
2506982  2019-12-31 19:26:48  2019-12-31 20:33:52               NaN   

         end_station_id           start_station_name         end_station_name  \
0                 196.0    Foothill Blvd at 42nd Ave  Grand Ave at Perkins St   
0                 196.0    Foothill Blvd at 42nd Ave  Grand Ave at Perkins St   
1                 134.0  Cyril Magnin St at Ellis St   Valencia St at 24th St   
2                 157.0       Downtown Berkeley BART     65th St at Hollis St   
3                  53.0      Church St at Duboce Ave   Grove St at Divisadero   
...                 ...                          ...                      ...   
2506978             NaN                          NaN                      NaN   
2506979             NaN                          NaN                      NaN   
2506980             NaN                          NaN                      NaN   
2506981             NaN                          NaN                      NaN   
2506982             NaN                          NaN                      NaN   

         bike_id   user_type       ts_start_hour  start_hour  start_weekday  \
0           4861  Subscriber 2019-01-31 17:00:00          17              3   
0           4861  Subscriber 2019-01-31 17:00:00          17              3   
1           5506  Subscriber 2019-01-31 20:00:00          20              3   
2           2717    Customer 2019-01-31 18:00:00          18              3   
3           4557    Customer 2019-01-31 19:00:00          19              3   
...          ...         ...                 ...         ...            ...   
2506978   719651  Subscriber 2019-12-28 18:00:00          18              5   
2506979   655411  Subscriber 2019-12-28 18:00:00          18              5   
2506980   936881  Subscriber 2019-12-26 14:00:00          14              3   
2506981   397539  Subscriber 2019-12-30 20:00:00          20              0   
2506982   662304  Subscriber 2019-12-31 19:00:00          19              1   

         start_month  trip_duration  start_station_latitude  \
0                  1        80825.0               37.775745   
0                  1        80825.0               37.774234   
1                  1        65901.0               37.785881   
2                  1        62634.0               37.870139   
3                  1        44681.0               37.770083   
...              ...            ...                     ...   
2506978           12          731.0                     NaN   
2506979           12          749.0                     NaN   
2506980           12         1045.0                     NaN   
2506981           12          326.0                     NaN   
2506982           12         4024.0                     NaN   

         start_station_longitude  
0                    -122.213037  
0                    -122.212483  
1                    -122.408915  
2                    -122.268422  
3                    -122.429156  
...                          ...  
2506978                      NaN  
2506979                      NaN  
2506980                      NaN  
2506981                      NaN  
2506982                      NaN  

[2618429 rows x 15 columns]

## Filter df on start_station coordinates

We found out that in the provided dataset, we have a lot of values that are actually not from San Francisco, but from Oakland or San José.  
This can be seen in the following map, which shows the station distribution.

In [20]:
#Control visualization that all data points are now only located in the city of San Francisco
mymap_start = folium.Map(location = [37.6, -122.2], 
                   width = 950, 
                   height = 550,
                   zoom_start = 10, 
                   tiles = 'openstreetmap')
                   
## show all the (start) stations
for lat, lng in zip(station_geo_data['start_station_latitude'],station_geo_data['start_station_longitude']):    
    station = folium.CircleMarker(
            location=[lat, lng],
            radius=3,
            color='red',
            fill=True,
            fill_color='yellow',
            fill_opacity=0.2)   
    # add the circle marker to the map
    station.add_to(mymap_start)


# show our selected area to filter our dataset
trail_coordinates = [
    (39, -122.32),
    (37.33, -122.32),
]
folium.PolyLine(trail_coordinates, tooltip="Coast").add_to(mymap_start)
trail_coordinates = [
    (37.33, -122.32),
    (37.33, -125),
]
folium.PolyLine(trail_coordinates, tooltip="Coast").add_to(mymap_start)

    
mymap_start

As we only want to analyze the data from San Francisco, we filter our dataset based on the start station coordinates.  
Everything that's below a latitiude of 37.33 and above a longitude of -122.32 will be kicked out.
Trips with empty start_station_ids are kept with the assumption that these trips happened within San Francisco.

In [21]:
bike_data_nan =  bike_data_joined[bike_data_joined.start_station_id.isna()]
bike_data_filtered = bike_data_joined[(bike_data_joined['start_station_latitude'] > 37.33 ) & (bike_data_joined['start_station_longitude'] < -122.32)]
bike_data_filtered = pd.concat([bike_data_filtered,bike_data_nan])
bike_data_filtered.head()

start_time             end_time  start_station_id  end_station_id  \
1  2019-01-31 20:58:34  2019-02-01 15:16:55               4.0           134.0   
3  2019-01-31 19:46:10  2019-02-01 08:10:51              85.0            53.0   
4  2019-01-31 14:19:02  2019-02-01 07:10:52              16.0            28.0   
6  2019-01-31 23:54:39  2019-02-01 00:14:28              34.0           146.0   
8  2019-01-31 23:17:06  2019-02-01 00:09:40              29.0            70.0   

               start_station_name              end_station_name  bike_id  \
1     Cyril Magnin St at Ellis St        Valencia St at 24th St     5506   
3         Church St at Duboce Ave        Grove St at Divisadero     4557   
4         Steuart St at Market St  The Embarcadero at Bryant St     2100   
6  Father Alfred E Boeddeker Park       30th St at San Jose Ave     5114   
8   O'Farrell St at Divisadero St        Central Ave at Fell St     4813   

    user_type       ts_start_hour  start_hour  start_weekday  start_month  \
1  Subscriber 2019-01-31 20:00:00          20              3            1   
3    Customer 2019-01-31 19:00:00          19              3            1   
4    Customer 2019-01-31 14:00:00          14              3            1   
6  Subscriber 2019-01-31 23:00:00          23              3            1   
8  Subscriber 2019-01-31 23:00:00          23              3            1   

   trip_duration  start_station_latitude  start_station_longitude  
1        65901.0               37.785881              -122.408915  
3        44681.0               37.770083              -122.429156  
4        60710.0               37.794130              -122.394430  
6         1189.0               37.783988              -122.412408  
8         3154.0               37.782405              -122.439446

We visualize our filtered stations once again:


In [22]:
# get df with geo coordinates data for each station
station_geo_data = bike_data_filtered[['start_station_id', 'start_station_latitude','start_station_longitude']].dropna().drop_duplicates()
station_geo_data.set_index('start_station_id', inplace=True)
station_geo_data

#Control visualization that all data points are now only located in the city of San Francisco
mymap_start = folium.Map(location = [37.6, -122.2], 
                   width = 950, 
                   height = 550,
                   zoom_start = 10, 
                   tiles = 'openstreetmap')

## show all the (start) stations
for lat, lng in zip(station_geo_data['start_station_latitude'],station_geo_data['start_station_longitude']):    
    station = folium.CircleMarker(
            location=[lat, lng],
            radius=3,
            color='red',
            fill=True,
            fill_color='yellow',
            fill_opacity=0.2)   
    # add the circle marker to the map
    station.add_to(mymap_start)

mymap_start

## Filter on outliners

We have some trips that are extremely long with a duration of several hours and even trips with negative durations.  
We decided to kick out all the negative values and also the highest 0.5 quantile regarding trip duration.

In [23]:
# filter negative trip_duration
bike_data_filtered = bike_data_filtered[bike_data_filtered.trip_duration >= 0]

In [24]:
print(f"Every trip that has a longer duration than {bike_data_filtered['trip_duration'].quantile(0.995) /60/60} hours is filtered out.")


Every trip that has a longer duration than 1.9491666666666667 hours is filtered out.


In [25]:
bike_data_filtered = bike_data_filtered[bike_data_filtered.trip_duration <=  bike_data_filtered.trip_duration.quantile(0.995)]

## Save prepared dataset

In [26]:
bike_data_filtered.to_csv("data/sf_2019_prepared.csv", index=False)